### PR TITLE
Change ToolPanel search to client side and Advanced search to backend

### PR DESCRIPTION
**Pull request by ahmedhamidawan**
The tool panel search used Whoosh backend search and filtered on a large number of fields which made it an expensive operation. In this PR, the following changes have been made: 
- The tool panel search has been limited to a simpler client side search that only filters on the `name+description` fields.
- The advanced tool search has been changed from client side filtering to a more comprehensive backend Whoosh search.

## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

### CODE REVIEW

File: client/src/components/Panels/Common/ToolSearch.vue [hunk 1]
```diff
@@ -82,24 +88,28 @@ export default {
     },
     methods: {
         checkQuery(q) {
+            this.filterSettings["name"] = q;
             this.$emit("onQuery", q);
             if (q && q.length >= this.minQueryLength) {
                 if (this.favorites.includes(q)) {
                     this.$emit("onResults", this.favoritesResults);
                 } else {
                     this.loading = true;
```
Comment (guerler) [inline client/src/components/Panels/Common/ToolSearch.vue:97]: Setting the loading flag here has no effect, removing it.

File: client/src/components/Panels/Common/ToolAdvancedSearchResults.vue [hunk 2]
```diff
@@ -107,14 +107,17 @@ export default {
...
     },
     methods: {
+        initialize() {
+            if (this.tools) {
+                this.tools.slice(this.bufferLen - defaultBufferLen, this.bufferLen).forEach(async (toolId) => {
+                    await this.fetchHelp(toolId);
+                });
+                this.busy = false;
+            }
+        },
```
Comment (guerler) [inline client/src/components/Panels/Common/ToolAdvancedSearchResults.vue]: Just fyi, @ahmedhamidawan and I talked offline and we will revise the data processing mechanism, which should significantly improve the overall performance.

File: client/src/store/toolStore.js [hunk 3]
```diff
@@ -57,13 +32,31 @@ const actions = {
...
+                // Do OR search on name+description field
+                if (key == "name") {
+                    q += "(" + key + ":(" + filterValue + ") OR description" + ":(" + filterValue + ")) ";
+                } else if (key == "id") {
+                    q += "id_exact:(" + filterValue + ") ";
+                } else {
+                    q += key + ":(" + filterValue + ") ";
+                }
+            }
+            return axios
```
Comment (guerler) [inline client/src/store/toolStore.js]: Suggesting to use `urlData` from `utils/url` here.

File: client/src/store/toolStore.js [hunk 4]
```diff
@@ -21,34 +22,8 @@ const getters = {
...
-                }
-                if (hasMatches) {
-                    returnedTools.push(allTools[tool]);
-                }
-            }
-
-            return returnedTools;
-        },
+        ({ toolbox }) =>
+            state.toolsList.length === 0 ? [] : filterTools(toolbox, state.toolsList),
```
Comment (guerler) [inline client/src/store/toolStore.js]: Should just return `filterTools`, the empty check should be in `filterTools`, and we could add a test for it if necessary.

File: client/src/store/toolStore.js [hunk 5]
```diff
@@ -57,13 +32,31 @@ const actions = {
...
-        // store isn't already populated (initial fetch)
-        if (state.toolsList.length === 0) {
-            console.log("fetching all tools once");
-            const { data } = await axios.get(`${getAppRoot()}api/tools?in_panel=False`);
-            commit("saveTools", { toolsData: data });
+    fetchAllTools: async ({ commit }, { filterSettings }) => {
+        if (Object.keys(filterSettings).length !== 0) {
+            // Parsing filterSettings to Whoosh query
+            let q = "";
+            for (const [key, filterValue] of Object.entries(filterSettings)) {
```
Comment (guerler) [inline client/src/store/toolStore.js]: Suggesting to push this into its own utility file and to add a test-function like e.g. here https://github.com/galaxyproject/galaxy/blob/dev/client/src/utils/url.test.js.

**Review by guerler — APPROVED** [review]
Great step into the right direction. It would be nice to follow-up with resolving some of the key-down event handlers i.e. pressing enter does not automatically trigger the search etc, but this can be done in follow-ups. Thank you so much for working on this.

**Comment by dannon** [comment]
This should resolve the last bits of what I was hoping for in a first pass for #14028, thank you!

**Pull request merged**